Readme.md
# woff2base64

Generates CSS font-face with woff2 and woff fonts embedded as base64.

[![Build Status](https://travis-ci.org/Fetten/woff2base64.svg?branch=master)](https://travis-ci.org/Fetten/woff2base64)

## Installation

```
npm install --save woff2base64
```

## Usage

woff2base64 automatically looks up the font-weight and font-style in the given font filename (case-insensitive). 
All [possible css values](https://developer.mozilla.org/en-US/docs/Web/CSS/font-weight) are supported.

If your font's filename contains the word _regular_, the font-weight will be set to `normal`.

However, if you cannot change the font's filename, you can pass the weight and style for each font via 
the `font` parameter (see below).

```js
import fs from 'fs';
import woff2base64 from 'woff2base64';
const fonts = {
    'Roboto-Bold.woff2': fs.readFileSync('fonts/Roboto/Roboto-Bold.woff2'),
    'Roboto-Bold.woff':  fs.readFileSync('fonts/Roboto/Roboto-Bold.woff'),
	// ...
};
const options = {
    fontFamily: 'Roboto'
};
const css = woff2base64(fonts, options);
// css.woff2 = '@font-face{font-family:"Roboto";src:url(data:application/font-woff2;charset=utf-8;base64,...'
// css.woff  = '@font-face{font-family:"Roboto";src:url(data:application/font-woff;charset=utf-8;base64,...'

fs.writeFileSync('build/roboto.woff2.css', css.woff2);
fs.writeFileSync('build/roboto.woff.css', css.woff);
```

## Parameters

woff2base64 accepts two objects: `fonts` and `options`.

### fonts (required)

Use woff2 and woff filenames as keys.

The value can either be a `String` containing the file content, 
a `Buffer` containing the file content or an `Object`, allowing you to pass font-specific options:

```js
const fonts = {
    'Roboto-BoldItalic.woff2': {
        content: fs.readFileSync('fonts/Roboto/Roboto-BoldItalic.woff2'),
        weight: 'bold',
        style: 'italic'
    },
    'Roboto-Bold.woff2': {
        content: fs.readFileSync('fonts/Roboto/Roboto-Bold.woff2'),
        weight: 'bold',
        style: 'normal'
    },
    'Roboto-Regular.woff2': fs.readFileSync('fonts/Roboto/Roboto-Bold.woff2'),
};
```

### Options

#### fontFamily (required)

The name of the Font Family used in css.

Type: `String`


#### fontFaceTemplate

The Template used to generate the css. 
Used with Lodash's [_.template](https://lodash.com/docs/#template).

Type: `String`

Default:
```
@font-face{ font-family:"<%=family%>";src:url(<%=uri%>) format("<%=format%>");font-weight:<%=weight%>;font-style:<%=style%>;}
```

#### fontUriTemplate

The Template used to generate the base64 encoded data uri. 
Used with Lodash's [_.template](https://lodash.com/docs/#template).

Type: `String`

Default:
```
data:<%=mime%>;charset=utf-8;base64,<%=base64%>
```

#### banner

A banner that gets prepended to generated css files.
Set to `''` to disable banner. 

Type: `String`


## Resources

* Convert your ttf fonts to woff2 and woff with [Font Squirrel’s Webfont Generator](http://www.fontsquirrel.com/tools/webfont-generator).

* See the current browser support for woff2 and woff fonts on [canisue.com](http://caniuse.com/#search=woff).

* Why would you need this? Read [Better webfont loading with using localStorage and providing WOFF2 support](http://bdadam.com/blog/better-webfont-loading-with-localstorage-and-woff2.html) 
by Adam Beres-Deak.

* This package is based on [fontoptim by Artem Sapegin](https://github.com/sapegin/fontoptim) with
improvements for font-style and font-weight recognition.

## License
Released under the [MIT license](License.md).

Copyright (c) 2017 Marcel Fetten. All rights reserved. 


Changelog.md
<a name="1.0.1"></a>
## [1.0.1](https://github.com/Fetten/woff2base64/compare/v1.0.0...v1.0.1) (2017-01-26)


### Bug Fixes

* **project:** remove support for node.js version < 5 ([0d2b24e](https://github.com/Fetten/woff2base64/commit/0d2b24e))



<a name="1.0.0"></a>
# 1.0.0 (2017-01-26)





test/test.js
'use strict';
import fs from 'fs';
import path from 'path';
import glob from 'glob';
import {chai, expect} from 'chai';
import Woff2Base64Class from '../lib/woff2base64';
import woff2base64 from '../lib/index';


let fonts = {};
let expected = {};
const options = {
  fontFamily: 'Roboto'
};

glob.sync('./test/fonts/Roboto/*.{woff,woff2}').forEach((name) => {
  fonts[path.basename(name)] = {content: fs.readFileSync(name)};
});

glob.sync('./test/expected/*.css').forEach((name) => {
  expected[path.basename(name, '.css')] = fs.readFileSync(name, {
    encoding: 'utf-8'
  });
});

describe('woff2base64', function () {
  describe('Woff2Base64 Class', function () {
    it('exists', function () {
      expect(Woff2Base64Class).to.exists;
    });

    it('accepts options', function () {
      const fo = new Woff2Base64Class({}, options);

      expect(fo.opts.fontFamily).to.equal('Roboto');
      expect(fo.opts.banner).to.equal('/* Generated by woff2base64 (https://github.com/fetten/woff2base64) */\n');
    });

    it('gets the font format from a filename', function () {
      const fo = new Woff2Base64Class({}, options);
      let format = fo.getFontFormat('Roboto-Regular.woff');
      expect(format).to.equal('woff');

      format = fo.getFontFormat('Roboto-Bold.woff2');
      expect(format).to.equal('woff2');
    });

    it('gets the font-weight from filename', function () {
      const fo = new Woff2Base64Class({}, options);
      let weight = fo.getFontWeightFromFilename('Roboto-Regular.woff');
      expect(weight).to.equal('normal');

      weight = fo.getFontWeightFromFilename('Roboto-Bold.woff2');
      expect(weight).to.equal('bold');

      weight = fo.getFontWeightFromFilename('Roboto-BlackItalic-900.woff2');
      expect(weight).to.equal('900');
    });

    it('gets the font-style from filename', function () {
      const fo = new Woff2Base64Class({}, options);

      let style = fo.getFontStyleFromFilename('Roboto-Bold.woff2');
      expect(style).to.equal('normal');

      style = fo.getFontStyleFromFilename('Roboto-Italic.woff2');
      expect(style).to.equal('italic');

      style = fo.getFontStyleFromFilename('Roboto-BlackItalic-900.woff2');
      expect(style).to.equal('italic');
    });

    it('converts strings to base64', function () {
      const fo = new Woff2Base64Class({}, options);
      const base64 = fo.stringToBase64('woff is awesome');

      expect(base64).to.equal('d29mZiBpcyBhd2Vzb21l');
    });

    it('gets the MimeType by font format', function () {
      const fo = new Woff2Base64Class({}, options);
      const woffMime = fo.getMimeType('woff');
      const woff2Mime = fo.getMimeType('woff2');

      expect(woffMime).to.equal('application/font-woff');
      expect(woff2Mime).to.equal('application/font-woff2');
    });

    it('gets the font data uri by string', function () {
      const fo = new Woff2Base64Class({}, options);
      const woffUri = fo.getFontDataUri('woff', 'woff is awesome');
      const woff2Uri = fo.getFontDataUri('woff2', 'woff2 is awesome');

      expect(woffUri).to.equal('data:application/font-woff;charset=utf-8;base64,d29mZiBpcyBhd2Vzb21l');
      expect(woff2Uri).to.equal('data:application/font-woff2;charset=utf-8;base64,d29mZjIgaXMgYXdlc29tZQ==');
    });

    it('gets the font data uri by Buffer', function () {
      const fo = new Woff2Base64Class({}, options);
      const woffUri = fo.getFontDataUri('woff', new Buffer('woff is awesome'));
      const woff2Uri = fo.getFontDataUri('woff2', new Buffer('woff2 is awesome'));

      expect(woffUri).to.equal('data:application/font-woff;charset=utf-8;base64,d29mZiBpcyBhd2Vzb21l');
      expect(woff2Uri).to.equal('data:application/font-woff2;charset=utf-8;base64,d29mZjIgaXMgYXdlc29tZQ==');
    });

    it('gets the FontFace css string', function () {
      const fo = new Woff2Base64Class({}, options);
      const woffFontFace = fo.getFontFace('Roboto-Bold.woff', 'woff', {content: 'Hello world'});
      const woff2FontFace = fo.getFontFace('Roboto-Bold.woff2', 'woff2', {content: 'Hello world'});

      expect(woffFontFace).to.equal(['@font-face{', 'font-family:"Roboto";', 'src:url(data:application/font-woff;charset=utf-8;base64,SGVsbG8gd29ybGQ=) format("woff");', 'font-weight:bold;', 'font-style:normal', '}'].join(''));
      expect(woff2FontFace).to.equal(['@font-face{', 'font-family:"Roboto";', 'src:url(data:application/font-woff2;charset=utf-8;base64,SGVsbG8gd29ybGQ=) format("woff2");', 'font-weight:bold;', 'font-style:normal', '}'].join(''));
    });

    it('generates an empty font list', function () {
      const fo = new Woff2Base64Class({}, options);
      const result = fo.generate();

      expect(result).to.be.an('object');
      expect(result).to.be.empty;
    });

    it('generates', function () {
      const fo = new Woff2Base64Class(fonts, options);
      const result = fo.generate();

      expect(result).to.be.an('object');
      expect(result).to.have.property('woff');
      expect(result).to.have.property('woff2');
      expect(result.woff).not.to.be.empty;
      expect(result.woff2).not.to.be.empty;

      expect(result.woff).to.equal(expected.woff);
      expect(result.woff2).to.equal(expected.woff2);
    });
  });

  describe('woff2base64 wrapper', function () {
    it('exists', function () {
      expect(woff2base64).to.exists;
    });

    it('generates', function () {
      const result = woff2base64(fonts, options);
      expect(result).to.be.an('object');
      expect(result).to.have.property('woff');
      expect(result).to.have.property('woff2');
      expect(result.woff).not.to.be.empty;
      expect(result.woff2).not.to.be.empty;

      expect(result.woff).to.equal(expected.woff);
      expect(result.woff2).to.equal(expected.woff2);
    });
  });

  describe('Readme example', function () {
    it('generates', function () {
      const example = require('./example');

      if (!fs.existsSync('./test/tmp')) {
        fs.mkdirSync('./test/tmp');
      }
      example();

      const woff = fs.readFileSync('./test/tmp/roboto.woff.css', {
        encoding: 'utf-8'
      });

      expect(woff).not.to.be.empty;
      expect(woff).to.equal(expected.woff);

      const woff2 = fs.readFileSync('./test/tmp/roboto.woff2.css', {
        encoding: 'utf-8'
      });

      expect(woff2).not.to.be.empty;
      expect(woff2).to.equal(expected.woff2);

      fs.unlinkSync('./test/tmp/roboto.woff.css');
      fs.unlinkSync('./test/tmp/roboto.woff2.css');
      fs.rmdirSync('./test/tmp');
    });
  });
});


test/example/index.js
module.exports = function () {
  'use strict';

  var fs = require('fs');
  var woff2base64 = require('../../lib/index');

  var fonts = {
    'Roboto-Black-900.woff': {
      content: fs.readFileSync('./test/fonts/Roboto/Roboto-Black-900.woff'),
      weight: '900',
      style: 'normal'
    },
    'Roboto-BlackItalic-900.woff': {
      content: fs.readFileSync('./test/fonts/Roboto/Roboto-BlackItalic-900.woff'),
      weight: '900',
      style: 'italic'
    },
    'Roboto-Bold.woff':             fs.readFileSync('./test/fonts/Roboto/Roboto-Bold.woff'),
    'Roboto-BoldItalic.woff':       fs.readFileSync('./test/fonts/Roboto/Roboto-BoldItalic.woff'),
    'Roboto-Italic.woff':           fs.readFileSync('./test/fonts/Roboto/Roboto-Italic.woff'),
    'Roboto-Light-300.woff':        fs.readFileSync('./test/fonts/Roboto/Roboto-Light-300.woff'),
    'Roboto-LightItalic-300.woff':  fs.readFileSync('./test/fonts/Roboto/Roboto-LightItalic-300.woff'),
    'Roboto-Medium-500.woff':       fs.readFileSync('./test/fonts/Roboto/Roboto-Medium-500.woff'),
    'Roboto-MediumItalic-500.woff': fs.readFileSync('./test/fonts/Roboto/Roboto-MediumItalic-500.woff'),
    'Roboto-Regular.woff':          fs.readFileSync('./test/fonts/Roboto/Roboto-Regular.woff'),
    'Roboto-Thin-100.woff':         fs.readFileSync('./test/fonts/Roboto/Roboto-Thin-100.woff'),
    'Roboto-ThinItalic-100.woff':   fs.readFileSync('./test/fonts/Roboto/Roboto-ThinItalic-100.woff'),

    'Roboto-Black-900.woff2': {
      content: fs.readFileSync('./test/fonts/Roboto/Roboto-Black-900.woff2'),
      weight: '900',
      style: 'normal'
    },
    'Roboto-BlackItalic-900.woff2': {
      content: fs.readFileSync('./test/fonts/Roboto/Roboto-BlackItalic-900.woff2'),
      weight: '900',
      style: 'italic'
    },
    'Roboto-Bold.woff2':              fs.readFileSync('./test/fonts/Roboto/Roboto-Bold.woff2'),
    'Roboto-BoldItalic.woff2':        fs.readFileSync('./test/fonts/Roboto/Roboto-BoldItalic.woff2'),
    'Roboto-Italic.woff2':            fs.readFileSync('./test/fonts/Roboto/Roboto-Italic.woff2'),
    'Roboto-Light-300.woff2':         fs.readFileSync('./test/fonts/Roboto/Roboto-Light-300.woff2'),
    'Roboto-LightItalic-300.woff2':   fs.readFileSync('./test/fonts/Roboto/Roboto-LightItalic-300.woff2'),
    'Roboto-Medium-500.woff2':        fs.readFileSync('./test/fonts/Roboto/Roboto-Medium-500.woff2'),
    'Roboto-MediumItalic-500.woff2':  fs.readFileSync('./test/fonts/Roboto/Roboto-MediumItalic-500.woff2'),
    'Roboto-Regular.woff2':           fs.readFileSync('./test/fonts/Roboto/Roboto-Regular.woff2'),
    'Roboto-Thin-100.woff2':          fs.readFileSync('./test/fonts/Roboto/Roboto-Thin-100.woff2'),
    'Roboto-ThinItalic-100.woff2':    fs.readFileSync('./test/fonts/Roboto/Roboto-ThinItalic-100.woff2'),
  };

  var css = woff2base64(fonts, {fontFamily: 'Roboto'});

  fs.writeFileSync('./test/tmp/roboto.woff.css', css.woff);
  fs.writeFileSync('./test/tmp/roboto.woff2.css', css.woff2);
};


src/woff2base64.js
/**
 * Woff2Base64 class.
 * Generates CSS with woff2 and woff fonts embedded as Base64.
 *
 * @copyright (c) 2017 Marcel Fetten (https://www.fetten-meier.com)
 * @license MIT
 */
'use strict';

import path from 'path';
import _ from 'lodash';

class Woff2Base64 {
  mimeTypes = {
    woff: 'application/font-woff',
    woff2: 'application/font-woff2'
  };

  fontWeights = [
    '100',
    '200',
    '300',
    '400',
    '500',
    '600',
    '700',
    '800',
    '900',
    'bolder',
    'bold',
    'lighter',
    'normal',
    'regular' // gets converted to `normal`
  ];

  fontStyles = [
    'italic',
    'oblique',
    'normal'
  ];

  defaultOpts = {
    fontFamily: null,
    fontFaceTemplate: [
      '@font-face{',
      'font-family:"<%=family%>";',
      'src:url(<%=uri%>) format("<%=format%>");',
      'font-weight:<%=weight%>;',
      'font-style:<%=style%>',
      '}'
    ].join(''),
    fontUriTemplate: 'data:<%=mime%>;charset=utf-8;base64,<%=base64%>',
    banner: '/* Generated by woff2base64 (https://github.com/fetten/woff2base64) */\n'
  };

  /**
   * Class Constructor
   *
   * @param {object} fonts Object of fonts. Filename as key, value can be a string or Buffer with the file content or an object.
   * @param {object} opts Options.
   */
  constructor(fonts, opts) {
    this.fonts = fonts;
    this.opts = {};

    _.extend(this.opts, this.defaultOpts, opts);
    this.opts.fontFaceTemplate = _.template(this.opts.fontFaceTemplate);
    this.opts.fontUriTemplate = _.template(this.opts.fontUriTemplate);

    if (!this.opts.fontFamily || typeof this.opts.fontFamily !== 'string') {
      throw new Error('No valid fontFamily given in options. Please provide a fontFamily (string).');
    }

    for (let filename in this.fonts) {
      if (!this.fonts.hasOwnProperty(filename)) {
        continue;
      }

      // font content can be passed as string or Buffer - convert it to object
      if (typeof (this.fonts[filename]) === 'string' || this.fonts[filename] instanceof Buffer) {
        this.fonts[filename] = {
          content: this.fonts[filename]
        };
      }
    }
  }

  /**
   * Generate
   * @return {object} Object with font-formats (e.g. `woff2` and `woff`) as keys and the css property as value.
   */
  generate = () => {
    let resultObject = this.getResultObject();
    resultObject = _.mapValues(resultObject, this.prependBannerAndJoinLines);

    return resultObject;
  };

  /**
   * Get the result object
   *
   * @returns {Object} Object with font-formats (e.g. `woff` and `woff2`) as keys and the css property as value.
   */
  getResultObject = () => {
    let formats = {};

    for (let filename in this.fonts) {
      if (!this.fonts.hasOwnProperty(filename)) {
        continue;
      }

      // skip any other filename than .woff or woff2
      if (!/.woff2?$/i.test(filename)) {
        continue;
      }

      let format = this.getFontFormat(filename);
      if (!formats[format]) {
        formats[format] = [];
      }

      let fontface = this.getFontFace(filename, format, this.fonts[filename]);
      formats[format].push(fontface);
    }

    return formats;
  };

  /**
   * Prepend the banner from options and join single lines with line-break
   *
   * @param lines
   * @returns {string}
   */
  prependBannerAndJoinLines = (lines) => {
    return this.opts.banner + lines.join('\n');
  };

  /**
   * Get the font-face css property from a given font
   *
   * @param {string} filename The font filename
   * @param {string} format The font format (woff2 or woff)
   * @param {object} font The font object
   * @returns {string} The css font-face property string
   */
  getFontFace = (filename, format, font) => {

    return this.opts.fontFaceTemplate({
      format,
      family: this.opts.fontFamily,
      uri: this.getFontDataUri(format, font.content),
      weight: this.getFontWeight(filename, font),
      style: this.getFontStyle(filename, font)
    });
  };

  /**
   * Get the base64-encoded data-uri of a font
   *
   * @param {string} format The font format (woff2 or woff)
   * @param {string|Buffer} fontContent The font-file content
   * @returns {string} Base64-encoded data-uri of a font
   */
  getFontDataUri = (format, fontContent) => {
    const base64 = this.stringToBase64(fontContent);
    const mime = this.getMimeType(format);

    return this.opts.fontUriTemplate({
      format,
      mime,
      base64
    });
  };

  /**
   * Get the mime-type for a font-format
   *
   * @param {string} format The font-format (woff2 or woff)
   * @returns {string} The mime-type for a font-format
   */
  getMimeType = (format) => {
    if (!this.mimeTypes.hasOwnProperty(format)) {
      throw new Error(`${format} is not a valid format.`);
    }
    else {
      return this.mimeTypes[format];
    }
  };

  /**
   * Get the font-format from a filename
   *
   * @param {string} filename The filename
   * @returns {string} The font-format
   */
  getFontFormat = (filename) => {
    const extension = path.extname(filename);

    return extension.substring(1);
  };


  /**
   * Encode a string to base64
   *
   * @param {String} string String to encode
   * @return {String} Base64 encoded string
   */
  stringToBase64 = (string) => {
    return new Buffer(string).toString('base64');
  };

  /**
   * Get the font-weight
   * Checks if a font-weight is present in the font-object or tries to guess the font-weight from the filename.
   *
   * @param {string} filename The font filename
   * @param {Object} font The font object
   * @returns {string} The font-weight. Defaults to `normal`.
   */
  getFontWeight = (filename, font) => {
    return font.hasOwnProperty('weight') &&
    this.fontWeights.hasOwnProperty(font.weight) ? font.weight : this.getFontWeightFromFilename(filename);

  };

  /**
   * Get the font-style
   * Checks if a font-style is present in the font-object or tries to guess the font-style from the filename.
   *
   * @param {string} filename The font filename
   * @param {Object} font The font object
   * @returns {string} The font-style. Defaults to `normal`.
   */
  getFontStyle = (filename, font) => {
    return font.hasOwnProperty('style') &&
    this.fontStyles.hasOwnProperty(font.style) ? font.style : this.getFontStyleFromFilename(filename);
  };

  /**
   * Guess the font-weight from the filename
   *
   * @param {string} filename The filename
   * @returns {string} The font-weight. Defaults to `normal`.
   */
  getFontWeightFromFilename = (filename) => {
    let weight = 'normal';

    for (let fontWeight in this.fontWeights) {
      if (!this.fontWeights.hasOwnProperty(fontWeight)) {
        continue;
      }

      let regex = new RegExp(`${this.fontWeights[fontWeight]}`, 'i');
      if (regex.test(filename)) {
        weight = this.fontWeights[fontWeight];
        break;
      }
    }

    // convert `regular` to `normal` css value
    if (weight.toLowerCase() === 'regular') {
      weight = 'normal';
    }

    return weight;
  };

  /**
   * Guess the font-style from the filename
   *
   * @param {string} filename The filename
   * @returns {string} The font-style. Defaults to `normal`.
   */
  getFontStyleFromFilename = (filename) => {
    let style = 'normal';

    for (let fontStyle in this.fontStyles) {
      if (!this.fontStyles.hasOwnProperty(fontStyle)) {
        continue;
      }

      let regex = new RegExp(`${this.fontStyles[fontStyle]}`, 'i');
      if (regex.test(filename)) {
        style = this.fontStyles[fontStyle];
        break;
      }
    }

    return style;
  };
}

module.exports = Woff2Base64;
